_posts/2019-04-24-accessing-kafka-part-2.md

```diff
@@ -0,0 +1,317 @@
+The other parts published so far are:_
+
+* _[Part 1 - Introduction](https://strimzi.io/2019/04/17/accessing-kafka-part-1.html)_
+* _Part 2 - Node Ports (this post)_
+
+# Node ports
+
+`NodePort` is a special type of [Kubernetes service](https://kubernetes.io/docs/concepts/services-networking/service/#nodeport).
+When such service is created, Kubernetes will allocated a port on all nodes of the Kubernetes cluster and will make sure that all traffic to this port is routed to the service and eventually to the pods behind this service.
+
+The routing of the traffic is done by the kube-proxy Kubernetes component.
+It doesn't matter on which node is your pod running.
+The node ports will be open on all nodes and the traffic will always reach your pod.
+So your clients need to connect to the node port on any of the nodes of the Kubernetes cluster and let Kubernetes handle the rest.
+
+The node port is by default selected from the port range 30000-32767.
+But this range can be changed in Kubernetes configuration.
+
+So, how do we use `NodePort` type services in Strimzi to expose Apache Kafka?
+
+# Exposing Kafka using node ports
+
+As a user, you can easily expose Kafka using node ports.
+All you need to do is to configure it in the Kafka custom resource.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta1
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    # ...
+    listeners:
+      # ...
+      external:
+        type: nodeport
+        tls: false
+    # ...
+```
+
+But what happens after you configure it is a bit more complicated.
+
+The first think we need to address is how will the clients access the individual brokers.
+As explained in the first part, having a service which will round robin across all brokers in the cluster will not work with Kafka.
+The clients need to be able to reach each of the brokers directly.
+Inside the Kubernetes cluster we addressed it by using the pod DNS names as the advertised addresses.
+But the pod hostnames or IP addresses are not recognized outside of Kubernetes, so we cannot use them.
+So, how does Strimzi solve it?
+
+Instead, of using the pod hostnames or IP addresses, we create additional services - one of each Kafka broker.
+So in a Kafka cluster with N brokers we will have N+1 node port services:
+
+* One which can be used by the Kafka clients as the bootstrap service for the initial connection and for receiving the metadata about the Kafka cluster
+* Another N services - one for each broker - to address the brokers directly
+
+All of these services are created with the type `NodePort`.
+Each of these services will be assigned different node port, so that the traffic for the different brokers can be distinguished.
+
+![Accessing Kafka using per-pod services]({{ "/assets/2019-04-24-per-pod-services.png" }})
+
+Since Kubernetes 1.9, every pod in a stateful set has automatically a label `statefulset.kubernetes.io/pod-name` which contains the name of the pod.
```

Comment:
```suggestion
Since Kubernetes 1.9, every pod in a stateful set is automatically labelled with `statefulset.kubernetes.io/pod-name` which contains the name of the pod.
```